lib/index.js

```diff
@@ -85,7 +85,7 @@ class ReactRefreshPlugin {
     }
 
     // Inject react-refresh context to all Webpack entry points
-    compiler.options.entry = injectRefreshEntry(compiler.options.entry, this.options);
+    compiler.options.entry = injectRefreshEntry(compiler.options.entry, this.options, compiler);
```

Comment:
I'm not quite certain about passing down the whole compiler object here, when what we only need is `compiler.optoins.entry` and `compiler.options.devServer` - can you consolidate the function to only accept `compiler.options` (i.e. `compilerOptions`) and `this.options` (i.e. `options`)?

Reply:
I can do that, but I think it will actually be quite a lot of work - `injectRefreshEntry` recursively calls itself in a lot of different places, and if we make this change then it would require all those recursive calls to to be wrapped in an object

for example, line 119 in `injectRefreshEntry.js` is 

```js
        injectRefreshEntry(resolvedEntry, options)
```

and we would probably have to change it to 

```js
        injectRefreshEntry({entry: resolvedEntry}, options)
```

in addition to updating all the tests in `injectRefreshEntry.test.js`. Again, I don't mind doing that, I just want to be sure that is what want to happen.

Alternatively, I could just change the third parameter to just taking in the `devServer` object and we wouldn't have to churn all the code/tests.

Let me know what you prefer.

Reply:
This is great point. Now that I look at the code, I realised that the options passed down are exclusively used for the overlay entry (and I think for your change you also need to pass the `compiler.options.devServer` object down to the recursion below to ensure that those entry types also get the correct metadata injected). We could probably do something a bit smarter - that is to extract out that part of the logic into some sort of a `getOverlayEntry` function, which does not need to recurse. That way we could keep the recursion as is without having to pollute it, and still ensuring the overlay entries and resource query is being correctly constructed.

What do you think?

Reply:
ah, hmm... where are you thinking that `getOverlayEntry` would be called? if we call it inside of `injectRefreshEntry` then we would still have to pass in the compiler stuff to `injectRefreshEntry` in the end, right? 

(I could be missing something though 🙂 )

Reply:
We could call it outside of `injectRefreshEntry` - like above the re-assign of `compiler.options.entry` - then pass the constructed error overlay entries so that theys would only be constructed once?

Reply:
ok I think I see where you're going! 😄 

I think I can get an updated version in 10 minutes or so, to see what you think

Reply:
ah but that would still require a rework of what is passed into `injectRefreshEntry` though, right? (and thus churn on the tests/recursive calls)

maybe I didn't fully get what you're saying

Reply:
> ah but that would still require a rework of what is passed into injectRefreshEntry though, right? (and thus churn on the tests/recursive calls)

Yes it would - mostly for tests - but I think this is a better way forward instead of augmenting the amount of parameters we pass into the recursion (and thus sometimes forgetting them) ...
And they are truly only needed for one pass ...

Reply:
Is this sort of what you were thinking? https://github.com/frehner/react-refresh-webpack-plugin/pull/2/files

note that I didn't update the tests yet because I wanted to be sure I was doing it right, before I went and changed all the tests.

Reply:
Left some review comments. Super thankful for you two to be driving this effort (and the one for `sockProtocol`)!